test/mobile/nnc/test_context.cpp

```diff
@@ -0,0 +1,156 @@
+#include <torch/csrc/jit/mobile/nnc/context.h>
+#include <torch/csrc/jit/mobile/nnc/registry.h>
+#include <ATen/Functions.h>
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace nnc {
+
+extern "C" {
+
+// out = a * n (doing calculation in the `tmp` buffer)
+int bad_mul_kernel(void** args) {
+  const int size = 128;
+  at::Tensor a = at::from_blob(args[0], {size}, at::kFloat);
+  at::Tensor out = at::from_blob(args[1], {size}, at::kFloat);
+  at::Tensor n = at::from_blob(args[2], {1}, at::kInt);
+  at::Tensor tmp = at::from_blob(args[3], {size}, at::kFloat);
+
+  tmp.zero_();
+  for (int i = n.item().toInt(); i > 0; i--) {
+    tmp.add_(a);
+  }
+  out.copy_(tmp);
+  return 0;
+}
+
+int dummy_kernel(void** /* args */) {
+  return 0;
+}
+
+} // extern "C"
+
+REGISTER_NNC_KERNEL("bad_mul", bad_mul_kernel)
+REGISTER_NNC_KERNEL("dummy", dummy_kernel)
+
+InputSpec create_test_input_spec(const std::vector<int64_t>& sizes) {
+  InputSpec input_spec;
+  input_spec.sizes_ = sizes;
+  input_spec.dtype_ = at::kFloat;
+  return input_spec;
+}
+
+OutputSpec create_test_output_spec(const std::vector<int64_t>& sizes) {
+  OutputSpec output_spec;
+  output_spec.sizes_ = sizes;
+  output_spec.dtype_ = at::kFloat;
+  return output_spec;
+}
+
+MemoryPlan create_test_memory_plan(const std::vector<int64_t>& buffer_sizes) {
+  MemoryPlan memory_plan;
+  memory_plan.buffer_sizes_ = buffer_sizes;
+  return memory_plan;
+}
+
+TEST(Function, ExecuteBadMul) {
+  const int a = 999;
+  const int n = 100;
+  const int size = 128;
+  Function f;
+
+  f.set_nnc_kernel_id("bad_mul");
+  f.set_input_specs({create_test_input_spec({size})});
+  f.set_output_spec({create_test_output_spec({size})});
+  f.set_parameters({at::ones({1}, at::kInt).mul(n)});
+  f.set_memory_plan(create_test_memory_plan({sizeof(float) * size}));
+
+  c10::List<at::Tensor> input({
+      at::ones({size}, at::kFloat).mul(a)
+  });
+  auto outputs = f.run(c10::impl::toList(input));
+  auto output = ((const c10::IValue&) outputs[0]).toTensor();
+  auto expected_output = at::ones({size}, at::kFloat).mul(a * n);
+  EXPECT_TRUE(output.equal(expected_output));
+}
+
+TEST(Function, Serialization) {
+  Function f;
+  f.set_name("test_function");
+  f.set_nnc_kernel_id("test_kernel");
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  f.set_input_specs({create_test_input_spec({1, 3, 224, 224})});
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  f.set_output_spec({create_test_output_spec({1000})});
+  f.set_parameters({
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({1, 16, 3, 3}, at::kFloat),
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({16, 32, 1, 1}, at::kFloat),
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+      at::ones({32, 1, 3, 3}, at::kFloat)
+  });
+  f.set_memory_plan(create_test_memory_plan({
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
```

Comment:
Yeah, I feel it will be cleaner if we have `Function`  with API `getScratchSpaceSizes` or something like that, which can be used with create_test_memory_plan.

Reply:
So "ScratchSpaceSizes" is about how to allocate temp scratch spaces, right? In my mental model I refer to this as "memory plan" (that's why I save it into MemoryPlan, and attach MemoryPlan to Function) - MemoryPlan is essentially off-spin of "ScratchSpaceSizes" from "Function", lol.